### PR TITLE
Charts: [Line Chart] Add better support for stacked series data

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/33349

## Background

| Before  | After |
| ------------- | ------------- |
| Screenshot 2022-04-04 at 14 56 38  | Screenshot 2022-04-04 at 14 56 54  |

As you can see prior to this PR the line chart had a bad visual and data stacking. If points from two or more series don't share one x-axis coordinate (time) the stacking logic fails and visually we didn't handle this as well, therefore wierd stacked area paths. 

In this PR we handle stacking in series data in the right way. If points from two series don't have the same x time coordinate we're trying to find the closest points from the series on the stack and interpolate values to find the base value for the current point. 

Screenshot 2022-04-04 at 15 06 02

## Test plan

- Make sure that all stories work as they worked before 
- Check the performance of the line chart rendering (should be the same or better)
- Check that stacked series work as they should for non-aligned by x-axis points

## App preview:
- [Link](https://sg-web-vk-fix-stacked-line-chart.onrender.com)

### CODE REVIEW

File: client/web/src/charts/components/line-chart/components/tooltip/TooltipContent.tsx [hunk 1]
```diff
@@ -17,38 +17,40 @@ export function getLineStroke<Datum>(line: Series<Datum>): string {
...
 /**
  * Display tooltip content for XYChart.
  * It consists of title - datetime for current x point and list of all nearest y points.
  */
 export function TooltipContent<Datum>(props: TooltipContentProps<Datum>): ReactElement | null {
-    const { activePoint, series, xAxisKey } = props
-    const { datum, originalDatum } = activePoint
+    const { activePoint, series, stacked } = props
+
+    console.log(props)
```
Comment (unclejustin) [inline client/web/src/charts/components/line-chart/components/tooltip/TooltipContent.tsx:35]:
```suggestion
```

File: client/web/src/charts/components/line-chart/utils/data-series-processing/get-series-data.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ *   { a: 2, b: 3, x: 2},
+ *   { a: 4, b: 5, x: 3},
+ *   { a: null, b: 6, x: 4}
+ * ] → series: [
+ *     { name: a, data: [{ y: 2, x: 2 }, { y: 4, x: 3 }],
+ *     { name: b, data: [{ y: 3, x: 2 }, { y: 5, x: 3 }, { y: 6, x: 4 }]
+ *   ]
+ * ```
+ */
+export function getSeriesData<Datum>(input: Input<Datum>): SeriesWithData<Datum>[] {
```
Comment (unclejustin) [inline client/web/src/charts/components/line-chart/utils/data-series-processing/get-series-data.ts:31]: Could you add a couple of unit tests for this function quickly? If so, I'd love to have some coverage.

File: client/web/src/charts/components/line-chart/utils/data-series-processing/get-stacked-series-data.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+ *   │                                       │  ●                        ●
+ *   │  ●                                    │  ▲     ●                  ▲
+ *   │        ●                  ●           │  │     ▲     ●     ●      │
+ *   │              ◇     ●      ◇           │  │     │     ◇     ▲      ◇
+ *   │        ◇                              │  │     ◇           │
+ *   │  ◇           ●     ◇                  │  ◇                 ◇
+ * ──┼──────────────────────────────▶      ──┼──────────────────────────────▶
+ * ```
+ */
+export function getStackedSeriesData<Datum>(input: Input<Datum>): StackedSeries<Datum>[] {
```
Comment (unclejustin) [inline client/web/src/charts/components/line-chart/utils/data-series-processing/get-stacked-series-data.ts:31]: Similar to above unit tests for these functions would be great if you can get them done without too much effort.

File: client/web/src/charts/components/line-chart/utils/data-series-processing/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * │░░░░░░░░░░░░▽░░    ■                │       │░░░░░░░░░░░░▽░░    ■                │
+ * │░░░░░░░░░░░░░░░          ▽          │──────▶│░░░░░░░░░░░░░░░                     │
+ * │░░░░░░▽░░░░░░░░ ■                   │       │░░░░░░▽░░░░░░░░ ■                   │
+ * │░░░░░░░░░░░░░░░       ■             │       │░░░░░░░░░░░░░░░       ■             │
+ * │░░░▽░░░░░░░░░░░                     │       │░░░▽░░░░░░░░░░░                     │
+ * │░░░░░░░░░░░░░░░             ▽       │       │░░░░░░░░░░░░░░░                     │
+ * └────────────────────────────────────┘       └────────────────────────────────────┘
+ *```
+ */
+export function getFilteredSeriesData<Datum>(data: SeriesDatum<Datum>[]): SeriesDatum<Datum>[] {
```
Comment (unclejustin) [inline client/web/src/charts/components/line-chart/utils/data-series-processing/helpers.ts:24]: Unit tests. I'm just flagging these now. I have an idea to cover all of them quickly that I will leave in a comment at the end.

File: client/web/src/charts/components/line-chart/utils/generate-points-field.ts [hunk 5]
```diff
@@ -1,43 +1,39 @@
...
-                          x: xScale(+datum[xAxisKey]),
-                          y: yScale(+datum[series.dataKey]),
-                          color: series.color ?? 'green',
-                          linkUrl: getLinkURL(datum),
-                          originalDatum: originalData[index],
-                          datum,
-                      }
-                    : null
-            )
+    const { dataSeries, xScale, yScale } = input
```
Comment (unclejustin) [inline client/web/src/charts/components/line-chart/utils/generate-points-field.ts:18]: Unit tests

File: client/web/src/charts/components/line-chart/utils/get-min-max-boundary.ts [hunk 6]
```diff
@@ -17,23 +17,27 @@ interface Boundaries {
  * (time interval) global for all lines on the chart.
  */
 export function getMinMaxBoundaries<D>(props: MinMaxBoundariesInput<D>): Boundaries {
-    const { dataSeries, xAxisKey } = props
+    const { dataSeries, zeroYAxisMin } = props
```
Comment (unclejustin) [inline client/web/src/charts/components/line-chart/utils/get-min-max-boundary.ts:20]: Unit tests.

**Review by unclejustin — APPROVED** [review]
Looks good to me 🌮 !

My main points were around unit test coverage for the utils functions. I think we can get good enough coverage for all these by just using `toMatchInlineSnapshot`.

Example:

```
expect(getMinMaxBoundaries(data)).toMatchInlineSnapshot()
```

That would at least give us coverage for the future if we ever change these functions.

**Comment by vovakulikov** [comment]
Thanks, @unclejustin. Yeah, I completely agree about those unit tests. I added tests for data calculation logic. UI logic will be handled by screenshot tests in chromatic

**Pull request merged**